### PR TITLE
Improved *NumberingScheme code for Forward detectors

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Moved from class to namespace. The code had no internal state so all the functions work fine as standalone functions.
- Converted C array use to std::vector. This avoids a warning from gcc12 and does automatic memory management.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by civanch** [comment]
+1

My impression that the solution in this PR a bit better than in https://github.com/cms-sw/cmssw/pull/41942 , full regression confirms.

File: SimG4CMS/Forward/src/BscNumberingScheme.cc [hunk 1]
```diff
@@ -11,93 +11,94 @@
...
 
-  LogDebug("BscSim") << "BscNumberingScheme number of levels= " << level;
+    LogDebug("BscSim") << "BscNumberingScheme number of levels= " << level;
 
-  if (level > 0) {
-    int* copyno = new int[level];
-    G4String* name = new G4String[level];
-    detectorLevel(aStep, level, copyno, name);
+    if (level > 0) {
+      std::vector<int> copyno;
```
Comment (VinInn) [inline SimG4CMS/Forward/src/BscNumberingScheme.cc:44]: why not just on the stack?
```
int copyno[level];
```
etc

Comment (civanch) [reply SimG4CMS/Forward/src/BscNumberingScheme.cc:44]: I think, compiler does not know what value may have "level".

Comment (Dr15Jones) [reply SimG4CMS/Forward/src/BscNumberingScheme.cc:44]: dynamic arrays are not part of the C++ standard so I do not use them.

Comment (VinInn) [reply SimG4CMS/Forward/src/BscNumberingScheme.cc:44]: still they are fully supported by all compilers and executable is faster

Comment (civanch) [reply SimG4CMS/Forward/src/BscNumberingScheme.cc:44]: These detectors rarely hit by particles, so overhead is minimal. In truth, it is possible to change this code in a way that both arrays and vectors are not used. The first attempt in https://github.com/cms-sw/cmssw/pull/41942 was not successful in testing. So, I would left things as they are.

**Comment by rappoccio** [comment]
+1

**Pull request merged**